### PR TITLE
Enable ASAN/LSAN/TSAN for *-apple-ios-macabi

### LINKED ISSUE #113935: Rust has no ASAN support for catalyst ABI (x86_64-apple-ios-macabi)
Building for `x86_64-apple-ios-macabi` with ASAN enabled produces this error:
```
 FAILED: local_rustc_sysroot/lib/rustlib/x86_64-apple-ios-macabi/lib/libcore.rlib 
 "python3" "../../build/rust/rustc_wrapper.py" --rustc=../../third_party/rust-toolchain/bin/rustc --depfile=local_rustc_sysroot/lib/rustlib/x86_64-apple-ios-macabi/lib/libcore.rlib.d --rsp=local_rustc_sysroot/lib/rustlib/x86_64-apple-ios-macabi/lib/libcore.rlib.rsp -- -Clinker="../../third_party/llvm-build/Release+Asserts/bin/clang++" --crate-name core ../../third_party/rust-toolchain/lib/rustlib/src/rust/library/core/src/lib.rs --crate-type rlib -Zforce-unstable-if-unmarked --edition=2021 -Coverflow-checks=on -Cdefault-linker-libraries -Zdep-info-omit-d-target -Zmacro-backtrace -Zremap-cwd-prefix=. --color=always --target=x86_64-apple-ios-macabi -Cembed-bitcode=no -Cpanic=abort -Zpanic_abort_tests --cfg cr_rustc_revision=\"006a26c0b546abc0fbef59a773639582b641e500-6-llvmorg-17-init-16420-g0c545a44\" -Zdwarf-version=4 -Cdebuginfo=1 -Zsanitizer=address --cap-lints=allow -Dunsafe_op_in_unsafe_fn --sysroot=local_rustc_sysroot   --emit=dep-info=local_rustc_sysroot/lib/rustlib/x86_64-apple-ios-macabi/lib/libcore.rlib.d,link -o local_rustc_sysroot/lib/rustlib/x86_64-apple-ios-macabi/lib/libcore.rlib LDFLAGS RUSTENV OUT_DIR=../../../../../../../../../out/Debug/gen/build/rust/std/rules/core CARGO_PKG_VERSION=0.0.0 CARGO_PKG_NAME=core CARGO_PKG_DESCRIPTION=The\ Rust\ Core\ Library
 error: address sanitizer is not supported for this target
```

The `aarch64-apple-ios-macabi` target gives the same compiler error.

Catalyst is a useful ABI for ASAN support because Chromium's iOS code is fuzzed on Catalyst, and ASAN is important to find bugs with fuzzers. This prevents us from having Rust code that is visible to Chrome's IOS product at all, or requires us to disable ASAN in the fuzzers.

Chromium issue: https://bugs.chromium.org/p/chromium/issues/detail?id=1466967

For someone coming in without background in Rust's sanitizer support, what would be needed to bring up ASAN support for this ABI? Consider that we don't need a runtime from the stdlib at all, as we're using Clang's ASAN runtime.

### PR DESCRIPTION
**Pull request by danakj**
The -macabi targets are iOS running on MacOS, and they use the runtime libraries for MacOS, thus they have the same sanitizers available as the *-apple-darwin targets.

This is based on the work of https://github.com/rust-lang/rust/commit/aacf3213b142f074999429eab767ef7b53c3a1a5.

Closes #113935.

### CODE REVIEW

**Comment by danakj** [comment]
Along the way to writing this CL, I added the targets to the `compiletest/src/util.rs` file but _didn't_ add them to the `spec/` files yet. At that point trying to build for asan still failed with the unsupported message (expected) but compiletests all passed. So I am not sure that this is being tested, any thoughts?

**Comment by danakj** [comment]
I have verified that building Chromium with Rust in the build, targeting ios-macabi with ASAN does link and run after this change.

**Comment by danakj** [comment]
> This PR changes how LLVM is built. Consider updating src/bootstrap/download-ci-llvm-stamp.

I think this may be a false positive, we're pointing to the asan libs and causing them to be built maybe instead of not when targeting `*-apple-ios-macabi`, but let me know if I am wrong and this does require touching the download-ci-llvm-stamp please.

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -574,7 +574,9 @@ fn copy_sanitizers(
             || target == "aarch64-apple-darwin"
             || target == "aarch64-apple-ios"
             || target == "aarch64-apple-ios-sim"
+            || target == "aarch64-apple-ios-catalyst"
```
Comment (thomcc) [inline src/bootstrap/compile.rs:577]: These don't look like the right target triples for Rust or LLVM. Shouldn't they be `$arch-apple-ios-macabi`?

**Review by thomcc — CHANGES_REQUESTED** [review]
Surprised this works at all. Shouldn't this mean the sanitizer runtimes aren't present and thus things won't load (and possibly even fail to link)? Actually, when this was tested, it was linked into chromium, so it's likely the chromium C++ build provided the sanitizer runtime.

After this is fixed, a smoke-test against pure-rust use should be done.

But in general I'm in favor of allowing this, so long as it works (once fixed and smoke tested). And like, it's an unstable feature on a tier 3 target too, so there's relatively low risk to accept.

File: src/bootstrap/compile.rs (hunk 1 above)
Comment (danakj) [reply src/bootstrap/compile.rs:577]: Oh yes, thanks, this is what we call the thing in chromium's build system and I typoed that in here.

Comment (danakj) [reply src/bootstrap/compile.rs]: This fact is reflected in the changes here: [src/bootstrap/llvm.rs](https://github.com/rust-lang/rust/pull/115644/files#diff-67761c315d694c57b071f47f7f2323d4260d7e161ea861511c5ee3e8c1d8e277)

The name of the runtime is "osx", not "iosmacabi" or something like that.

**Comment by danakj** [comment]
Bump for review?

**Review by thomcc — APPROVED** [review]
Looks fine to me. Not a `compiler/*` reviewer so I won't r+ it.

**Comment by cjgillot** [comment]
Thanks for the review @thomcc.
@bors r=cjgillot,thomcc

**Comment by danakj** [comment]
Thank you so much. Rust for iOS chromium is coming up. 😊

**Comment by danakj** [comment]
> 422 Update is not a fast forward

Hm I tried rebasing but there's no conflicts. I will upload the rebase anyhow.

**Pull request merged**